File: Tests/RedlineCoreTests/SidecarTests.swift
// The published sidecar. It is a contract with tools this project does not control, so the
// spellings other readers rely on are asserted rather than assumed.
import XCTest
@testable import RedlineCore

final class SidecarTests: XCTestCase {
    private var dir: URL!
    private let now = Date(timeIntervalSince1970: 1_760_000_000)

    override func setUpWithError() throws {
        dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("redline-sidecar-\(UUID().uuidString)")
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    private var windows: [LimitWindow] {
        [
            LimitWindow(provider: "Claude", key: "five_hour", utilization: 42,
                        resetsAt: now.addingTimeInterval(3600), source: .official),
            LimitWindow(provider: "Claude", key: "seven_day", utilization: 18,
                        resetsAt: now.addingTimeInterval(86400), source: .official),
            LimitWindow(provider: "Claude", key: "seven_day_opus", utilization: 71,
                        resetsAt: now.addingTimeInterval(86400), source: .official),
            LimitWindow(provider: "Codex", key: "seven_day", utilization: 5,
                        resetsAt: nil, source: .official),
        ]
    }

    func testPayloadCarriesBothPercentageSpellings() throws {
        let json = Sidecar.payload(windows: windows, updatedAt: now, producer: "redline/test")
        let five = try XCTUnwrap(json["five_hour"] as? [String: Any])
        XCTAssertEqual(five["used_percentage"] as? Double, 42)
        XCTAssertEqual(five["utilization"] as? Double, 42)
        XCTAssertNotNil(five["resets_at"])
    }

    func testModelScopedWindowsKeepTheirDisplayName() throws {
        let json = Sidecar.payload(windows: windows, updatedAt: now, producer: "redline/test")
        let scoped = try XCTUnwrap(json["model_scoped"] as? [[String: Any]])
        XCTAssertEqual(scoped.count, 1)
        XCTAssertEqual(scoped[0]["display_name"] as? String, "Opus")
        XCTAssertEqual(scoped[0]["utilization"] as? Double, 71)
    }

    func testOtherProvidersStayOutOfTheStandardKeys() throws {
        let json = Sidecar.payload(windows: windows, updatedAt: now, producer: "redline/test")
        // The standard block is Claude-shaped; Codex would be misread as Claude's week
        let seven = try XCTUnwrap(json["seven_day"] as? [String: Any])
        XCTAssertEqual(seven["utilization"] as? Double, 18)
        let extra = try XCTUnwrap(json["redline"] as? [String: Any])
        let all = try XCTUnwrap(extra["windows"] as? [[String: Any]])
        XCTAssertEqual(all.count, 4, "every provider still travels in the namespaced block")
    }

    func testRoundTripsThroughOurOwnParser() throws {
        let url = dir.appendingPathComponent("usage-snapshot.json")
        XCTAssertTrue(Sidecar.publish(windows: windows, updatedAt: now,
                                      producer: "redline/test", to: url))
        let parsed = try XCTUnwrap(StatuslineFeed.read(path: url, now: now))
        let stamp = try XCTUnwrap(parsed.updatedAt)
        XCTAssertEqual(stamp.timeIntervalSince1970, now.timeIntervalSince1970, accuracy: 1)
        XCTAssertEqual(parsed.windows.count, 3)
        XCTAssertTrue(parsed.isFresh(now: now))
    }

    func testPublishedFileIsPrivate() throws {
        let url = dir.appendingPathComponent("usage-snapshot.json")
        Sidecar.publish(windows: windows, updatedAt: now, producer: "redline/test", to: url)
        let perms = try FileManager.default
            .attributesOfItem(atPath: url.path)[.posixPermissions] as? NSNumber
        XCTAssertEqual(perms?.int16Value, 0o600)
    }

    func testExternalPathMustBeAbsoluteJSON() {
        XCTAssertNil(Sidecar.validExternalPath("usage.json"))
        XCTAssertNil(Sidecar.validExternalPath(""))
        #if os(Windows)
        // Absolute means drive-qualified or UNC here. A bare leading slash is rooted on
        // whichever drive happens to be current, which is not a path worth trusting.
        XCTAssertNil(Sidecar.validExternalPath(#"C:\tmp\usage.txt"#))
        XCTAssertNil(Sidecar.validExternalPath("/tmp/usage.json"))
        XCTAssertNotNil(Sidecar.validExternalPath(#"C:\tmp\usage.json"#))
        XCTAssertNotNil(Sidecar.validExternalPath(#"\\host\share\usage.json"#))
        #else
        XCTAssertNil(Sidecar.validExternalPath("/tmp/usage.txt"))
        XCTAssertNotNil(Sidecar.validExternalPath("/tmp/usage.json"))
        #endif
    }

    func testStaleExternalSidecarIsIgnored() throws {
        let url = dir.appendingPathComponent("other.json")
        Sidecar.publish(windows: windows, updatedAt: now.addingTimeInterval(-7200),
                        producer: "other/1.0", to: url)
        XCTAssertNil(Sidecar.readExternal(path: url.path, now: now),
                     "a file another tool stopped updating must not stand in for a live one")
        XCTAssertNotNil(Sidecar.readExternal(path: url.path,
                                             now: now.addingTimeInterval(-7200)))
    }

    func testEmptySidecarIsNotASource() throws {
        let url = dir.appendingPathComponent("empty.json")
        try Data("{\"updated_at\":\"2026-08-18T00:00:00Z\"}".utf8).write(to: url)
        XCTAssertNil(Sidecar.readExternal(path: url.path,
                                          now: ISO8601DateFormatter()
                                            .date(from: "2026-08-18T00:01:00Z")!))
    }
}

File: windows/tests/Redline.Core.Tests/SidecarTests.cs
// The published sidecar. It is a contract with tools this project does not control, so the
// spellings other readers rely on are asserted rather than assumed.
using System.Text.Json.Nodes;
using Redline.Core;

namespace Redline.Core.Tests;

public sealed class SidecarTests : IDisposable
{
    private readonly string _dir;
    private readonly DateTimeOffset _now = DateTimeOffset.FromUnixTimeSeconds(1_760_000_000);

    public SidecarTests()
    {
        _dir = Path.Combine(Path.GetTempPath(), "redline-sidecar-" + Guid.NewGuid());
        Directory.CreateDirectory(_dir);
    }

    public void Dispose()
    {
        try { Directory.Delete(_dir, true); } catch { }
    }

    private List<LimitWindow> Windows => new()
    {
        new("Claude", "five_hour", 42, _now.AddSeconds(3600), Provenance.Official),
        new("Claude", "seven_day", 18, _now.AddSeconds(86400), Provenance.Official),
        new("Claude", "seven_day_opus", 71, _now.AddSeconds(86400), Provenance.Official),
        new("Codex", "seven_day", 5, null, Provenance.Official),
    };

    [Fact]
    public void PayloadCarriesBothPercentageSpellings()
    {
        var json = Sidecar.Payload(Windows, _now, "redline/test");
        var five = Assert.IsType<JsonObject>(json["five_hour"]);
        Assert.Equal(42, Json.Num(five["used_percentage"]));
        Assert.Equal(42, Json.Num(five["utilization"]));
        Assert.NotNull(five["resets_at"]);
    }

    [Fact]
    public void ModelScopedWindowsKeepTheirDisplayName()
    {
        var json = Sidecar.Payload(Windows, _now, "redline/test");
        var scoped = Assert.IsType<JsonArray>(json["model_scoped"]);
        Assert.Single(scoped);
        Assert.Equal("Opus", Json.Str(scoped[0]!["display_name"]));
        Assert.Equal(71, Json.Num(scoped[0]!["utilization"]));
    }

    [Fact]
    public void OtherProvidersStayOutOfTheStandardKeys()
    {
        var json = Sidecar.Payload(Windows, _now, "redline/test");
        // The standard block is Claude-shaped; Codex would be misread as Claude's week
        var seven = Assert.IsType<JsonObject>(json["seven_day"]);
        Assert.Equal(18, Json.Num(seven["utilization"]));
        var extra = Assert.IsType<JsonObject>(json["redline"]);
        var all = Assert.IsType<JsonArray>(extra["windows"]);
        // Every provider still travels in the namespaced block
        Assert.Equal(4, all.Count);
    }

    [Fact]
    public void RoundTripsThroughOurOwnParser()
    {
        var path = Path.Combine(_dir, "usage-snapshot.json");
        Assert.True(Sidecar.Publish(Windows, "redline/test", updatedAt: _now, to: path));
        var parsed = StatuslineFeed.Read(path, _now);
        Assert.NotNull(parsed);
        var stamp = Assert.NotNull(parsed!.UpdatedAt);
        Assert.InRange((stamp - _now).TotalSeconds, -1, 1);
        Assert.Equal(3, parsed.Windows.Count);
        Assert.True(parsed.IsFresh(_now));
    }

    // chmod 0600 has no Windows equivalent (the profile is already private), so this pins
    // what remains: the write is atomic and leaves no temp file beside the sidecar.
    [Fact]
    public void PublishedFileIsPrivate()
    {
        var path = Path.Combine(_dir, "usage-snapshot.json");
        Assert.True(Sidecar.Publish(Windows, "redline/test", updatedAt: _now, to: path));
        Assert.True(File.Exists(path));
        Assert.Single(Directory.GetFiles(_dir));
    }

    [Fact]
    public void ExternalPathMustBeAbsoluteJSON()
    {
        Assert.Null(Sidecar.ValidExternalPath("usage.json"));
        Assert.Null(Sidecar.ValidExternalPath(@"C:\tmp\usage.txt"));
        Assert.Null(Sidecar.ValidExternalPath("/tmp/usage.txt"));
        Assert.Null(Sidecar.ValidExternalPath(""));
        Assert.NotNull(Sidecar.ValidExternalPath(@"C:\tmp\usage.json"));
        Assert.NotNull(Sidecar.ValidExternalPath("~/usage.json"));
    }

    [Fact]
    public void StaleExternalSidecarIsIgnored()
    {
        var path = Path.Combine(_dir, "other.json");
        Sidecar.Publish(Windows, "other/1.0", updatedAt: _now.AddSeconds(-7200), to: path);
        // A file another tool stopped updating must not stand in for a live one
        Assert.Null(Sidecar.ReadExternal(path, now: _now));
        Assert.NotNull(Sidecar.ReadExternal(path, now: _now.AddSeconds(-7200)));
    }

    [Fact]
    public void EmptySidecarIsNotASource()
    {
        var path = Path.Combine(_dir, "empty.json");
        File.WriteAllText(path, "{\"updated_at\":\"2026-08-18T00:00:00Z\"}");
        Assert.Null(Sidecar.ReadExternal(path, now: DateTimeOffset.Parse("2026-08-18T00:01:00Z")));
    }
}
